File: 99-Continue/159/DepthBinaryTree.py
from collections import deque

class Node:
    def __init__(self,data:int):
        self.left = self.right = None
        self.data = data
        
class BinaryTree:
    def __init__(self):
        self.root = None
        
    def buildTree(self,arr:list[str])->Node:
        if len(arr)==0:
            return self.root
        
        self.root=Node(int(arr[0]))
        queue=deque()
        queue.append(self.root)
        i = 1;
        while queue and i<len(arr):
            current = queue.popleft()
            if i<len(arr):
                if arr[i]!="null":
                    current.left = Node(int(arr[i]))
                    queue.append(current.left)
                i+=1
            if i<len(arr):
                if arr[i]!="null":
                    current.right = Node(int(arr[i]))
                    queue.append(current.right)
                i+=1
        return self.root
    
    def depthFinder(self)->int:
        if self.root is None:
            return 0
        
        depth = 0
        queue = deque([self.root])
        
        while queue:
            size = len(queue)
            while size>0:
                current = queue.popleft()
                if current.left is not None:
                    queue.append(current.left)
                if current.right is not None:
                    queue.append(current.right)
                size-=1
            depth+=1
        
        return depth
        
def main():
    tree = BinaryTree()
    n = int(input())
    lst = input().split()
    tree.buildTree(lst)
    print(tree.depthFinder())

if __name__ == "__main__":
    main()
    
                    
        
            
        

File: 99-Continue/159/DepthBinaryTree.java
import java.util.*;

public class DepthBinaryTree {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        BinaryTree tree = new BinaryTree();
        int n = sc.nextInt();
        sc.nextLine();
        String s = sc.nextLine();
        tree.buildTree(s.split(" "),sc);
        System.out.println(tree.depthFinder());
        sc.close();
        
    }
}

class BinaryTree{
    Node root;
    BinaryTree(){
        this.root = null;
    }
    
    Node buildTree(String[] arr,Scanner sc){
        if(arr.length==0 || arr[0].equals("null")){
            return root;
        }
        root = new Node(Integer.parseInt(arr[0]));
        int i=1;
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        
        while(!queue.isEmpty()&&i<arr.length){
            Node current = queue.poll();
            if(i<arr.length){
                if(!arr[i].equals("null")){
                    current.left = new Node(Integer.parseInt(arr[i]));
                    queue.offer(current.left);
                }
                i++;
            }
            if(i<arr.length){
                if(!arr[i].equals("null")){
                    current.right = new Node(Integer.parseInt(arr[i]));
                    queue.offer(current.right);
                }
                i++;
            }
        }
        
        return root;
    }
    
    int depthFinder() {
    if (root == null)
        return 0;

    Queue<Node> queue = new LinkedList<>();
    queue.offer(root);

    int depth = 0;

    while (!queue.isEmpty()) {

        int size = queue.size();

        while (size-- > 0) {
            Node current = queue.poll();

            if (current.left != null)
                queue.offer(current.left);

            if (current.right != null)
                queue.offer(current.right);
        }

        depth++;
    }

    return depth;
}
}

class Node{
    int data;
    Node left, right;
    Node(int data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}